core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java

```diff
@@ -174,6 +175,19 @@ public AbstractRuleBuilder onException() {
         return onException((unused1, unused2) -> true);
     }
 
+    /**
+     * Adds {@link TimeoutException}.
+     */
+    public AbstractRuleBuilder onTimeoutException() {
+        return onException((ctx, ex) -> {
+            if (ctx.isTimedOut()) {
+                return true;
+            }
+            return ex instanceof UnprocessedRequestException &&
+                   ex.getCause() instanceof TimeoutException;
```

Comment:
Shouldn't we do this? If users manually complete the response exceptionally with a `ResponseTimeoutException` from a decorator, it wouldn't work.
This might also confuse users that `onTimeoutException()` is invoked only when the exception is wrapped with the `UnprocessedRequestException` even though it isn't.

Reply:
> If users manually complete the response exceptionally with a ResponseTimeoutException from a decorator, it wouldn't work.

oh~ this case can happen. @ikhoon nim, WDYT? (how about check `ex instanceof TimeoutException` too?)

Reply:
> If users manually complete the response exceptionally with a ResponseTimeoutException from a decorator, it wouldn't work.

Hmm... `ResponseTimeoutException` is a special exception to cancel a response based on the timeout stored in `ctx.responseTimeoutMillis()`. If a user raises `ResponseTimeoutException`, it should be assumed as a normal exception but neither a timeout exception or a cancellation exception.

> This might also confuse users that onTimeoutException() is invoked only when the exception is wrapped with the UnprocessedRequestException even though it isn't.

If we update the contract of `ctx.isTimedOut()` properly, I don't see users get confused of the behavior. They may not judge the behavior by the implementation.

Anyway, it is good to add `ex instanceof TimeoutException` in this PR because `ctx.isTimedOut()` doesn't cover `WriteTimeoutException` and `ConcurrencyLimitTimeoutException` even though we will handle them in the following PR.
Let's make this PR solid itself.

Reply:
> Anyway, it is good to add ex instanceof TimeoutException in this PR

✅ updated!